server/channels/store/storetest/channel_store_categories.go

```diff
@@ -2413,3 +2423,251 @@ func testSidebarCategoryDeadlock(t *testing.T, rctx request.CTX, ss store.Store)
 
 	wg.Wait()
 }
+
+func testSidebarCategoryConcurrentAccess(t *testing.T, rctx request.CTX, ss store.Store, s SqlStore) {
+	if s.DriverName() == model.DatabaseDriverMysql {
+		t.Skip("This is known to fail on MySQL")
+	}
+
+	userID := model.NewId()
+	team := setupTeam(t, rctx, ss, userID)
+
+	numGoroutines := 40
+
+	// Create regular channels
+	channels := make([]*model.Channel, 5)
+	for i := 0; i < len(channels); i++ {
+		channel, nErr := ss.Channel().Save(rctx, &model.Channel{
+			Name:        fmt.Sprintf("channel-%d", i),
+			DisplayName: fmt.Sprintf("Channel %d", i),
+			Type:        model.ChannelTypeOpen,
+			TeamId:      team.Id,
+		}, 10)
+		require.NoError(t, nErr)
+		_, err := ss.Channel().SaveMember(rctx, &model.ChannelMember{
+			UserId:      userID,
+			ChannelId:   channel.Id,
+			NotifyProps: model.GetDefaultChannelNotifyProps(),
+		})
+		require.NoError(t, err)
+		channels[i] = channel
+	}
+
+	// Create DM channels to exercise different code paths
+	dmChannels := make([]*model.Channel, 3)
+	for i := 0; i < len(dmChannels); i++ {
+		otherUserID := model.NewId()
+		dmChannel, nErr := ss.Channel().CreateDirectChannel(rctx, &model.User{
+			Id: userID,
+		}, &model.User{
+			Id: otherUserID,
+		})
+		require.NoError(t, nErr)
+		dmChannels[i] = dmChannel
+	}
+
+	// Put them into a channel so that we can ensure they're evenly used across goroutines, and put 4x the number of
+	// goroutines in since that should ensure we have enough
+	channelChan := make(chan string, 4*numGoroutines)
+	for i := 0; i < cap(channelChan); i++ {
+		channelChan <- channels[i%len(channels)].Id
+	}
+	dmChannelChan := make(chan string, 4*numGoroutines)
+	for i := 0; i < cap(dmChannelChan); i++ {
+		dmChannelChan <- dmChannels[i%len(dmChannels)].Id
+	}
+
+	// Create the initial categories
+	opts := &store.SidebarCategorySearchOpts{
+		TeamID:      team.Id,
+		ExcludeTeam: false,
+	}
+	res, nErr := ss.Channel().CreateInitialSidebarCategories(rctx, userID, opts)
+	require.NoError(t, nErr)
+	require.NotEmpty(t, res)
+
+	initialCategories, err := ss.Channel().GetSidebarCategoriesForTeamForUser(userID, team.Id)
+	require.NoError(t, err)
+
+	// Create custom categories
+	customCategories := make([]*model.SidebarCategoryWithChannels, 2)
+	for i := 0; i < 2; i++ {
+		customCategory, err := ss.Channel().CreateSidebarCategory(userID, team.Id, &model.SidebarCategoryWithChannels{
+			SidebarCategory: model.SidebarCategory{
+				DisplayName: fmt.Sprintf("Custom Category %d", i),
+			},
+		})
+		require.NoError(t, err)
+		customCategories[i] = customCategory
+	}
+
+	// Run concurrent operations
+	var wg sync.WaitGroup
+
+	println("-----------------------", userID, team.Id)
```

Comment:
Debug statement. Let's remove this :)

Reply:
Huh. I didn't know you could write `println` without using the `fmt` package. I'll remove that